src/mlpack/methods/neat/neat_impl.hpp

```diff
@@ -0,0 +1,669 @@
+    NEAT(TaskType& task,
+         const size_t inputNodeCount,
+         const size_t outputNodeCount,
+         const size_t popSize,
+         const size_t maxGen,
+         const size_t numSpecies,
+         const double initialBias,
+         const double initialWeight,
+         const double weightMutationProb,
+         const double weightMutationSize,
+         const double biasMutationProb,
+         const double biasMutationSize,
+         const double nodeAdditionProb,
+         const double connAdditionProb,
+         const double connDeletionProb,
+         const double disableProb,
+         const double elitismProp,
+         const double finalFitness,
+         const size_t complexityThreshold,
+         const size_t maxSimplifyGen,
+         const bool isAcyclic):
+    task(task),
+    inputNodeCount(inputNodeCount),
+    outputNodeCount(outputNodeCount),
+    popSize(popSize),
+    maxGen(maxGen),
+    numSpecies(numSpecies),
+    initialBias(initialBias),
+    initialWeight(initialWeight),
+    weightMutationProb(weightMutationProb),
+    weightMutationSize(weightMutationSize),
+    biasMutationProb(biasMutationProb),
+    biasMutationSize(biasMutationSize),
+    nodeAdditionProb(nodeAdditionProb),
+    connAdditionProb(connAdditionProb),
+    connDeletionProb(connDeletionProb),
+    disableProb(disableProb),
+    elitismProp(elitismProp),
+    finalFitness(finalFitness),
+    complexityThreshold(complexityThreshold),
+    currentComplexityCeiling(complexityThreshold),
+    meanComplexity(0),
+    lastTransitionGen(0),
+    maxSimplifyGen(maxSimplifyGen),
+    searchMode(0),
+    isAcyclic(isAcyclic),
+    contenderNum(0),
+    tournamentSelectProb(0)
+{ /* Nothing to do here yet */ }
+
+// The main loop of the NEAT algorithm. Returns the best genome.
+template <class TaskType,
+          class ActivationFunction,
+          class SelectionPolicy>
+Genome<ActivationFunction> NEAT<TaskType, ActivationFunction, SelectionPolicy>
+    ::Train()
+{
+  Genome<ActivationFunction>::nextInnovID = 0;
+  Genome<ActivationFunction>::mutationBuffer.clear();
+
+  // Initialize.
+  Initialize();
+  speciesList = std::vector<std::vector<Genome<ActivationFunction>>>
+      (numSpecies);
+  Speciate(true);
+
+  // Main loop.
+  for (size_t gen = 0; gen < maxGen; gen++)
+  {
+    meanComplexity = 0;
+    arma::vec fitnesses(popSize);
+
+    #pragma omp parallel for
+    for (size_t i = 0; i < popSize; i++)
+    {
+      if (complexityThreshold != 0)
+        meanComplexity += (double)genomeList[i].Complexity() / popSize;
+      genomeList[i].Fitness() = task.Evaluate(genomeList[i]);
+      fitnesses[i] = genomeList[i].Fitness();
+    }
+
+    // Check termination criteria.
+    if (finalFitness != 0 && fitnesses.max() >= finalFitness)
+      break;
+    else if (gen == maxGen - 1)
+      break;
+
+    // Set search mode.
+    if (meanComplexity > currentComplexityCeiling)
+    {
+      if (searchMode == 0)
```

Comment:
I'm wondering if we should implement the search mode as a policy, so that we could add more methods in the future, without touching the NEAT implementation. Let me know what you think.

Reply:
I had though of this, but I couldn't come up with a general policy implementation since I hadn't heard of any other search types. Moreover in this case it's achievable with just changing probabilities, but in other cases a search mode could require a different kind of mutation, new rules and measures (I'm just theorizing here, but phased searching involved keeping track of complexity metrics and adding a method to measure complexity) and we would have to change the NEAT implementation anyway. Is there a way I could come up with some kind of general policy representation?